### DISCUSSION TITLE
Can I use 5 mpu6050 in the DMP example?

### DISCUSSION THREAD

**Discussion by schosdas** [jrowberg/i2cdevlib#663] [Q&A]
Hello
I want to connect 5 mpu6050 to Arduino and output roll, pitch, and yaw values. For this, I'm going to use the DMP_V6 example.
I know I can change the address value of the mpu6050, but can I use five sensors at the same time? If possible, can you give us a simple example?
Thank you.

**Comment by ZHomeSlice** [comment]
@schosdas 
The answer is yes it is possible to use multiple MPU's at the same time all with DMP V6.0. While I am sure it is possible to use Jeff's MPU6050 library I developed a similar and effective MPU6050 library that uses a slightly different approach to programming. So with that said you could possibly adapt my example code to use Jeffs MPU6050 Library or you are free to use my [Simple_MPU6050 library](https://github.com/ZHomeSlice/Simple_MPU6050)
Example Code:
[4 MPU 6050's at the same time](https://github.com/ZHomeSlice/Simple_MPU6050/blob/master/Examples/Simple_MPU6050_4_MPUs_Example/Simple_MPU6050_4_MPUs_Example.ino): 

How it works:
It will modify the DA0 pin to change the address of each MPU6050 on the fly only one of the mpu6050's will have the default address while the others will have the alternate address. then the sketch will cycle through each to upload firmware and manage the data.

Z

**Comment by jrowberg** [comment]
Each MPU-6050 can use one of two addresses based on the AD0 pin, which means on any given single I2C bus, you can have two MPUs. In order to run five different MPUs simultaneously, you would need to have three separate I2C buses, and/or an I2C bus multiplexer (which might be simpler in this case).

The only other possibility--which I'm not sure will work--is to toggle the AD0 pin on each IMU in a round-robin way so that the "active" one is set to GND (0x68 address) and all others are set to VDD (0x69 address), and communicate with each one in turn. This will fail if the MPU doesn't support switching its AD0 detection in real time, i.e. between different transactions. I've never tried it before, so I'm not sure.

**Comment by ZHomeSlice** [comment]
@jrowberg 

”The only other possibility--which I'm not sure will work--is to toggle the AD0 pin on each IMU in a round-robin way so that the "active" one is set to GND (0x68 address) and all others are set to VDD (0x69 address), and communicate with each one in turn. This will fail if the MPU doesn't support switching its AD0 detection in real time, i.e. between different transactions. I've never tried it before, so I'm not sure."

This method works great!
I've done 4 at a time with no issues.
Adding a fifth should not be a problem.
The limiting factor is getting the data across the i2c bus in time to catch the next packet.

**Comment by inhardwetrust** [comment]
Hi there,
I'm writing here cause I cant make a new topic (dont know why)...

In Arduino lib (MPU6050.h) we have following descriptions of bits(of MPU6050_RA_USER_CTRL        0x6A).

#define MPU6050_USERCTRL_DMP_EN_BIT             7
#define MPU6050_USERCTRL_FIFO_EN_BIT            6
#define MPU6050_USERCTRL_I2C_MST_EN_BIT         5
#define MPU6050_USERCTRL_I2C_IF_DIS_BIT         4
#define MPU6050_USERCTRL_DMP_RESET_BIT          3
#define MPU6050_USERCTRL_FIFO_RESET_BIT         2
#define MPU6050_USERCTRL_I2C_MST_RESET_BIT      1
#define MPU6050_USERCTRL_SIG_COND_RESET_BIT     0
![image](https://user-images.githubusercontent.com/108121689/175500152-253a11b5-24cc-4834-bbee-7becbf18219e.png)


But official Manual 
https://invensense.tdk.com/wp-content/uploads/2015/02/MPU-6000-Register-Map1.pdf
says that "Bits 7 and 3 are reserved" (page 38)
![image](https://user-images.githubusercontent.com/108121689/175500042-69544ae5-156e-4bf0-8f12-92c3685b791b.png)


Do you have some other Register Map Manual where this things are documented?

Kind Regards,
V

**Comment by jrowberg** [comment]
In short, no. Unfortunately.

Nearly all of the DMP-related implementation came from reverse-engineering the way the original InvenSense ARM-based evaluation driver board worked with the MPU-6050 plug-in daughterboard.

**Reply by ZHomeSlice** [reply]
@inhardwetrust 
The DMP library for versions 5.1.3 is not used here and version 6.12 (The only difference is in the DMP Image between 5.1.3 and 6.12) has details about these reserved bits in the documentation for implementing the DMP firmware

Here's my notes for my version simple_mpu6050:

```
#define USER_CTRL_WRITE_DMP_EN(Data)                    MPUi2cWrite(0x6A, 1, 7, (uint8_t)Data)		//   1  Enable DMP operation mode.
#define USER_CTRL_WRITE_FIFO_EN(Data)                   MPUi2cWrite(0x6A, 1, 6, (uint8_t)Data)		//   1  Enable FIFO operation mode. 0  Disable FIFO access from serial interface.  To disable FIFO writes by dma, use FIFO_EN register. To disable possible FIFO writes from DMP, disable the DMP.
#define USER_CTRL_WRITE_I2C_MST_EN(Data)                MPUi2cWrite(0x6A, 1, 5, (uint8_t)Data)		//   1  Enable the I2C Master (BIT_AUX_IF_EN)
#define USER_CTRL_WRITE_I2C_IF_DIS(Data)                MPUi2cWrite(0x6A, 1, 4, (uint8_t)Data)		//   1  Disable I2C Slave module and put the serial interface in SPI mode only.
#define USER_CTRL_WRITE_DMP_RST(...)                    MPUi2cWrite(0x6A, 1, 3, (uint8_t)1)			//   Reset DMP module. Reset is asynchronous. This bit auto clears after one clock cycle.
#define USER_CTRL_WRITE_FIFO_RST(...)                   MPUi2cWrite(0x6A, 1, 2, (uint8_t)1)			//   Reset FIFO module. Reset is asynchronous. This bit auto clears after one clock cycle.
#define USER_CTRL_WRITE_I2C_MST_RESET_BIT(...)          MPUi2cWrite(0x6A, 1, 1, (uint8_t)1)			//   Reset  gyro digital signal path, accel digital signal path, and temp digital signal path. This bit also clears all the sensor registers. SIG_COND_RST is a pulse of one clk8M wide.
#define USER_CTRL_WRITE_SIG_COND_RST(...)               MPUi2cWrite(0x6A, 1, 0, (uint8_t)1)			//   Reset  gyro digital signal path, accel digital signal path, and temp digital signal path. This bit also clears all the sensor registers. SIG_COND_RST is a pulse of one clk8M wide.
```
my code notes that describe these bits.

I'll see if I can find the detailed documentation that I dug this out of. I'll probably have to post an image of the PDF page. 

The documentation I mentioned is downloadable from the manufacturer's site.
Z

PDF Document by InvenSense:
Programming Sequence for DMP Hardware
Page 39

> Functions
> 9) Start Using DMP Hardware Features
> Once the MPU and DMP have been set up and configured to use DMP hardware features, the DMP is ready
> to process data and output the results into the FIFO.
> To start using the DMP hardware features, please perform the following steps:
> 9.1) Reset and Enable FIFO & DMP:
> 1. Write {0x40} bit 6 to USER_CTRL (0x6A) 
> 2. Write {0x04} bit 2 to USER_CTRL (0x6A)
> 3. Write {0x80} bit 7 to USER_CTRL (0x6A)
> 4. Write {0x08} bit 3 to USER_CTRL (0x6A)
> 9.2) Enable DMP Interrupt
> 5. Write {0x02} bit 1 to INT_ENABLE (0x38)
> Interrupts will be sent from the DMP at the configured output rate.
Note that the bits auto-clear after one clock cycle. 

I believe there are more details but I'm still looking.

**Reply by Alphanlipi** [reply]
Can I get the raw values using dmp in mpu6050 without using libraries

**Reply by jrowberg** [reply]
> Can I get the raw values using dmp in mpu6050 without using libraries

You can get raw values for acceleration (including gravity) and rotational velocity directly from the registers, but the DMP (orientation, gravity-less linear acceleration, etc.) is not accessible without a much more involved library. You have to load a blob of binary microcode in to make the DMP run in the first place, and then work with the FIFO for the output data.

**Comment by inhardwetrust** [comment]
Thanks! I got it
it looks I need to find a time to make it for STM32...

Just found this guy who made example for AVR without motion axis
https://github.com/bzerk/MPU6050_DMP_6_axis_demo_/blob/master/MPU6050_DMP_6_axis_demo_.pde

Now I need to port it to STM - to get rid of Wire library
